frontend/src/app/utils/dates/timeline.helper.ts

```diff
@@ -0,0 +1,155 @@
+import { addDays, addHours, addMonths, addWeeks, addYears, daysInWeek, differenceInHours, differenceInYears, endOfDay, endOfMonth, endOfWeek, endOfYear, startOfDay, startOfMonth, startOfWeek, startOfYear } from 'date-fns';
+
+import { DatePipe } from '@angular/common';
+import { TimelineScale } from '#types/tasks/timeline/enums/timeline-scale';
+import { TimelineSection } from '#types/tasks/timeline/timeline-section';
+
+export const getTimelineBoundaries = (currentDate: Date, scale: TimelineScale, pageNumber: number): TimelineSection => {
+    switch (scale) {
+    case TimelineScale.Day:
+        return {
+            start: startOfDay(addDays(currentDate, -1 * pageNumber / 3)),
+            end: endOfDay(addDays(currentDate, pageNumber / 3))
+        };
+    case TimelineScale.Week:
+        return {
+            start: startOfWeek(addWeeks(currentDate, -1 * pageNumber / 2)),
+            end: endOfWeek(addWeeks(currentDate, pageNumber / 2))
+        };
+    case TimelineScale.Month:
+        return {
+            start: startOfMonth(addMonths(currentDate, -1 * pageNumber / 2)),
+            end: endOfMonth(addMonths(currentDate, pageNumber / 2))
+        };
+    case TimelineScale.Year:
+        return {
+            start: startOfMonth(addYears(currentDate, -1 * pageNumber / 2)),
+            end: endOfMonth(addYears(currentDate, pageNumber / 2))
+        };
+    default:
+        throw new Error('Invalid scale');
+    }
+};
+
+export const getTimelineSections = (boundaries: TimelineSection, scale: TimelineScale) => {
+    if (scale !== TimelineScale.Year) {
+        return getTimelineSubsections(boundaries, scale + 1);
+    }
+    let sectionHours = 24 * 367;
+    let sectionsCount = differenceInYears(boundaries.end, boundaries.start);
+    let sections: TimelineSection[] = [];
+    let firstSectionStart = startOfYear(boundaries.start);
+    sections.push({ start: firstSectionStart, end: endOfYear(firstSectionStart) });
+    for (let i = 1; i < sectionsCount; i++) {
```

Comment:
I'm not sure, but wouldn't it be more clear to use do...while?
Actually all these loops in this class and timeline component a little bit hard to understand
also, not clear separation - some methods create data there, some of them here

I guess you need single separate timeline creator (it may include different services for different parts of timeline creation). So it would look like in timeline.component you would calll timelineCreator.create(...) with all required parameters and defaults. And it will return object that containts your model to apply on template. We may discuss it on a call.

Reply:
Separated data creation from the component